irohad/ordering/impl/on_demand_connection_manager.cpp

```diff
@@ -5,37 +5,52 @@
 
 #include "ordering/impl/on_demand_connection_manager.hpp"
 
+#include <chrono>
+
 #include <boost/range/combine.hpp>
 #include "interfaces/iroha_internal/proposal.hpp"
 #include "ordering/impl/on_demand_common.hpp"
 
+/// The time to wait for the up-to-date propagation data arrives through
+/// propagation_params_observable (see constructor) when we got a request for a
+/// round that is newer than our propagation data
+static constexpr std::chrono::milliseconds kPropagationDataWaitTimeout(10);
+
 using namespace iroha::ordering;
 
 OnDemandConnectionManager::OnDemandConnectionManager(
     std::shared_ptr<transport::OdOsNotificationFactory> factory,
-    rxcpp::observable<CurrentPeers> peers)
+    rxcpp::observable<PropagationParams> propagation_params_observable)
     : log_(logger::log("OnDemandConnectionManager")),
       factory_(std::move(factory)),
-      subscription_(peers.subscribe([this](const auto &peers) {
-        // exclusive lock
-        std::lock_guard<std::shared_timed_mutex> lock(mutex_);
+      subscription_(propagation_params_observable.subscribe(
+          [this](const auto &propagation_params) {
+            // exclusive lock
+            std::unique_lock<std::shared_timed_mutex> lock(mutex_);
 
-        this->initializeConnections(peers);
-      })) {}
+            this->initializeConnections(propagation_params);
+          })) {}
 
 OnDemandConnectionManager::OnDemandConnectionManager(
     std::shared_ptr<transport::OdOsNotificationFactory> factory,
-    rxcpp::observable<CurrentPeers> peers,
-    CurrentPeers initial_peers)
-    : OnDemandConnectionManager(std::move(factory), peers) {
-  // using start_with(initial_peers) results in deadlock
-  initializeConnections(initial_peers);
+    rxcpp::observable<PropagationParams> propagation_params_observable,
+    PropagationParams inital_propagation_params)
+    : OnDemandConnectionManager(std::move(factory),
+                                propagation_params_observable) {
+  // using start_with(inital_propagation_params) results in deadlock
+  initializeConnections(inital_propagation_params);
 }
 
-void OnDemandConnectionManager::onBatches(consensus::Round round,
+void OnDemandConnectionManager::onBatches(const consensus::Round round,
                                           CollectionType batches) {
   std::shared_lock<std::shared_timed_mutex> lock(mutex_);
 
+  // check if we have an appropriate propagation data:
+  if (!verifyPropagtionData(round, lock)) {
```

Comment:
It's hard to follow the function logic. I'd suggest passing waiting callback instead, e.g:

```cpp
verifyPropagationData(round, propagation_data_update_cv_.wait_for(
          lock, kPropagationDataWaitTimeout, [this, &request_round] {
            return not(this->current_round_ < request_round);
          }));
```

Reply:
or maybe even remove function and inline it's body here